internal/repository/repository.go

```diff
@@ -723,6 +723,21 @@ func (r *Repository) prepareCache() error {
 		fmt.Fprintf(os.Stderr, "error clearing pack files in cache: %v\n", err)
 	}
 
+	// clear old snapshots
+	snapshots := restic.NewIDSet()
+	err = r.List(context.TODO(), restic.SnapshotFile, func(id restic.ID, _ int64) error {
+		snapshots.Insert(id)
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error listing snapshots: %v\n", err)
+	}
+
+	err = r.Cache.Clear(restic.SnapshotFile, snapshots)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "error clearing snapshot files in cache: %v\n", err)
+	}
```

Comment:
I realised I can do the same thing, where ever we are listing the snapshots, as mentioned in the original issue.

Reply:
Explicitly listing the snapshots is somewhat cleaner, but requires an extra List call that must be avoided.

I can imagine two implementation variants: either hijack Repository.List or maybe better add a List method to the cache backend and let it take care of cleaning up indexes and snapshots itself. The latter variant has the benefit of keeping most of the cache management within the cache itself.